cherry_picker/cherry_picker.py

```diff
@@ -149,14 +151,20 @@ def remember_previous_branch(self):
     @property
     def upstream(self):
         """Get the remote name to use for upstream branches
-        Uses "upstream" if it exists, "origin" otherwise
+
+        Uses the remote passed to `--upstream-remote`.
+        If this flag wasn't passed, it uses "upstream" if it exists or "origin" otherwise.
         """
+        if self.upstream_remote is not None:
```

Comment:
I definitely disagree about backing down to `"upstream"`, I would say that telling the user that they've done something wrong (passed non-existent remote name) is better than silently falling back to a different remote name.

When it comes to `"origin"` though, it would make some amount of sense since this would make it so `self.upstream_remote` is just a replacement for `"upstream"` so basically (probably would just make `"upstream"` as the default value of `--upstream-remote` to simplify this):
```py
    @property
    def upstream(self):
        upstream_name = self.upstream_remote or "upstream"
        cmd = ["git", "remote", "get-url", upstream_name]
        try:
            self.run_cmd(cmd)
        except subprocess.CalledProcessError:
            self.upstream_remote = "origin"
        else:
            self.upstream_remote = upstream_name
        return self.upstream_remote
```

I'm unsure whether this is configurable enough. If it isn't, I could instead make `--upstream-remote` be renamed to `--upstream-remotes` and have it act as a list of remote names to try in order (with the default list being `["upstream", "origin"]`):
```py
    @property
    def upstream(self):
        # check the remote calculation cache
        if self.upstream_remote is not None:
            return self.upstream_remote

        for remote_name in self.upstream_remotes:
            cmd = ["git", "remote", "get-url", upstream_name]
            try:
                self.run_cmd(cmd)
            except subprocess.CalledProcessError:
                pass
            else:
                self.upstream_remote = upstream_name
                return upstream_name

        raise SomeKindOfError(...)
```

This might be a bit too complex though. I like the idea of having `"origin"` as fallback the most (and more than what's currently done in the PR) but I'd love to hear another opinion, ideally from someone who is interested in this feature.

@markshannon Since you're the one that made this feature request, what do you think?

Reply:
I should also add that *if* the feature request for configuring user preferences (#37) gets accepted, this would be easily overridable per repository without requiring the use of an option flag in such a scenario which is probably all the more reason not to implement the more complex solution with a list of remote names to try rather than just a single remote name.

Reply:
I think it would be better to check for the specified upstream remote and bail early with a meaningful message if the specified remote doesn't exist.  If no remote it's specified, we can keep checking for `"upstream"` and fallback on `"origin"` if `"upstream"` is missing.